pkg/cloud/api/mutable_resource.go

```diff
@@ -272,24 +291,51 @@ func (u *mutableResource[GA, Alpha, Beta]) AccessBeta(f func(x *Beta)) error {
 	return u.postAccess(meta.VersionBeta, 0)
 }
 
+// completeType means that the object does not have conversion errors and its
+// type is not missing.
+func completeType(err error, obj interface{}) bool {
+	return err == nil && !IsTypeMissing(obj)
+}
+
+// convertibleTo means that the object does not have conversion errors or its
+// type is missing.
+func convertibleTo(err error, obj interface{}) bool {
+	return err == nil || IsTypeMissing(obj)
+}
+
+// convertibleTo means that the object has conversion errors or its type is
+// missing.
+func notConvertibleTo(err error, obj interface{}) bool {
+	return err != nil || IsTypeMissing(obj)
+}
+
```

Comment:
We should document the version conversion logic now that it is more complex due to PlaceholderType

```
// ImpliedVersion returns the implied version of the underlying resource.
// This is determined by the convertibility of the resource.
//
//  y = Can convert to this version.
//  n = Can't convert (note: nothing can convert to PlaceholderType)
//
//  | GA| A | B |
//  -------------
//  | y | y | y | GA
//  | y | y | n | Alpha
//  | y | n | y | Beta
//  | y | n | n | GA (Note: this is somewhat weird as GA /should/ convert)
//  | n | y | y | Error (can't determine version)
//  | n | y | n | Alpha
//  | n | n | y | Beta
//  | n | n | n | Error (can't determine version)
```

Reply:
//  | y | n | n | GA (Note: this is somewhat weird as GA /should/ convert)
-- I think we should clearly said that the only case where we will imply GA version and conversion Alpha and Beta fails is where those types are `PlaceholderType` in other cases we should return error.
Also according to previous code we will deduce Alpha version only when GA and Beta conversion fails so
`| y | y | n | Alpha` and `| y | n | y | Beta` are not true

Proposition 
//  y = Can convert to this version. 
//  n = Can't convert (note: conversion of PlaceholderType will return 
// UseOfPlaceholderTypeError. This is a special case not covered here)
//  y/p = can convert or conversion returned UseOfPlaceholderTypeError
//  p = conversion returned UseOfPlaceholderTypeError

//  | GA| A | B |
//  -------------
//  | y | y/p | y/p | GA
//  | n | y   | n   | Alpha
//  | n | n   | y   | Beta
//  | y | n   | n   | Error (can't determine version)
//  | n | n   | n   | Error (can't determine version)
//  | p | y/n | y/n | Error (GA cannot be PlaceholderType)

Reply:
the discussion was moved offline. Resolving this thread